### ISSUE TITLE
memory leak

### ISSUE THREAD

**Issue by rpedela** [restify/node-restify#431]
I am pretty sure I found a major memory leak. It seems to be related to the `gzipResponse` and `bodyParser` plugins. I use [node-memwatch](https://github.com/lloyd/node-memwatch) to inspect the memory usage and force a garbage collection.

**Version Info**

``` js
ubuntu: 12.04
node: 0.10.4
restify: 2.6.0
```

**Server**

``` js
var fs = require('fs');
var memwatch = require('memwatch');
var restify = require('restify');

var hd = null;

function startDiff(req, res, next) {
    memwatch.gc();
    hd = new memwatch.HeapDiff();
    return next();
}

function endDiff(req, res, next) {
    memwatch.gc();
    var diff = hd.end();

    console.log('before: ' + diff.before.size);
    console.log('after: ' + diff.after.size);
    console.log('change: ' + diff.change.size);
    console.log('---------------------');

    return next();
}

var server = restify.createServer();

// removing either plugin also exhibits the memory leak
// remove both plugins and no memory leak
server.use(restify.gzipResponse());
server.use(restify.bodyParser());

server.get('/', [
    startDiff,
    function(req, res, next) {
        // large-file.json is a just a large JSON file, content does not matter
        var data = fs.readFileSync('large-file.json'));
        res.send(JSON.parse(data));
        return next();
    },
    endDiff,
]);

server.listen(4001, function() {
    console.log('listening at ' + server.url);
});
```

**Client**

```
curl --compressed http://localhost:4001 > /dev/null
```

**Output after 3 client calls, with memory leak**

```
before: 5.99 mb
after: 175.48 mb
change: 169.5 mb
---------------------
before: 175.83 mb
after: 318.21 mb
change: 142.38 mb
---------------------
before: 318.4 mb
after: 460.75 mb
change: 142.35 mb
---------------------
```

**Output after 3 client calls, with no memory leak**

```
before: 5.87 mb
after: 147.82 mb
change: 141.94 mb
---------------------
before: 5.36 mb
after: 147.86 mb
change: 142.5 mb
---------------------
before: 5.4 mb
after: 147.88 mb
change: 142.48 mb
---------------------
```

**Comment by micahr** [comment]
Haven't heard anything further regarding a potential memory leak in the Gzip Response plugin. Please re-open if this is still an issue.

**Comment by rpedela** [comment]
While old, I don't think it is appropriate to just close this issue. Memory leaks are one of those bugs/issues that are assumed a problem until proven otherwise. Just because no one has said anything, is not proof that the leak is fixed. In this case, it is really easy to see if the memory leak still exists. Run my test code above against the latest version and see if it still leaks.

**Comment by yunong** [comment]
Re-opened and assigned to myself. I'll take a look for the upcoming release.

**Comment by micahr** [comment]
Thank you, @rpedela . We'll take a look and see if the issue still exists. I apologize for closing the issue without checking if it was still a problem.

**Comment by DonutEspresso** [comment]
I'm not seeing this happen in node 4.x. Using the example code above (and migrating to memwatch-next):

```sh
$ node gzip.js
listening at http://[::]:4001
before: 7.16 mb
after: 32.18 mb
change: 25.02 mb
---------------------
before: 7.35 mb
after: 32.35 mb
change: 25 mb
---------------------
before: 7.36 mb
after: 32.36 mb
change: 25 mb
---------------------
before: 7.36 mb
after: 32.36 mb
change: 25 mb
---------------------
before: 7.37 mb
after: 32.37 mb
change: 25 mb
---------------------
before: 7.38 mb
after: 32.38 mb
change: 25 mb
---------------------
```

**Comment by weekens** [comment]
I had similar issue on version 2.7.0. The issue was resolved by upgrading to 5.0.0-beta1.

**Comment by retrohacker** [comment]
This sounds like it is resolved in v5 of node running on Node.js v4 and greater.

If you are still running into this, please comment and we can re-visit! :heart:
